include/http/middleware/prerender.js

```diff
@@ -0,0 +1,204 @@
+const url = require('url');
+const request = require('request-promise');
+const zlib = require('zlib');
+
+// This is based on https://github.com/RisingStack/koa-prerender
+module.exports = pb => ({
+    prerender: async (req, res) => {
+        try {
+            const enablePrerender = pb.config && pb.config.flags && pb.config.flags.enablePrerender;
+            const siteObj = pb.RequestHandler.sites[req.handler.hostname];
+            const enableSitePrerender = siteObj && siteObj.enableSitePrerender;
+
+            if (!enablePrerender && !enableSitePrerender) return;
+
+            const option = {
+                userAgent: req.headers['user-agent'],
+                bufferAgent: req.headers['x-bufferbot'],
+                method: req.method,
+                url: req.url
+            };
+
+            if (!shouldPreRender(option)) return;
+
+            const content = await getPrerenderedPage(req);
+            await renderResponse(req, content);
+
+        } catch(ex) {
+            // Just go normally path if there is any errors happen.
+            console.log(ex);
+        }
+    }
+});
+
+const extensionsToIgnore = [
+    '.js',
+    '.css',
+    '.xml',
+    '.less',
+    '.png',
+    '.jpg',
+    '.jpeg',
+    '.gif',
+    '.pdf',
+    '.doc',
+    '.txt',
+    '.ico',
+    '.rss',
+    '.zip',
+    '.mp3',
+    '.rar',
+    '.exe',
+    '.wmv',
+    '.doc',
+    '.avi',
+    '.ppt',
+    '.mpg',
+    '.mpeg',
+    '.tif',
+    '.wav',
+    '.mov',
+    '.psd',
+    '.ai',
+    '.xls',
+    '.mp4',
+    '.m4a',
+    '.swf',
+    '.dat',
+    '.dmg',
+    '.iso',
+    '.flv',
+    '.m4v',
+    '.torrent'
+];
+
+const crawlerUserAgents = [
```

Comment:
minor: move it out to separate file then require it here, I believe now the file is too long